src/api/v4/error-handling/carto-error.js

```diff
@@ -8,25 +24,51 @@ var UNEXPECTED_ERROR = 'unexpected error';
  *
  * @api
  */
-function CartoError (error) {
+function CartoError (error, opts) {
+  opts = opts || {};
+  var cartoError = _.extend({
+    message: UNEXPECTED_ERROR,
+    type: '',
+    origin: GENERIC_ORIGIN
+  }, error);
+
   if (_isWindshaftError(error)) {
-    this.message = error.errors[0];
-    return this;
-  }
-  if (error && error.message) {
-    this.message = error.message;
-    return this;
+    cartoError = transformWindshaftError(error, opts.layers, opts.analysis);
   }
   if (error && error.responseText) {
-    this.message = _handleAjaxResponse(error);
-    return this;
+    cartoError.message = _handleAjaxResponse(error);
+    cartoError.origin = 'ajax';
+    cartoError.type = error.statusText;
   }
 
-  this.message = UNEXPECTED_ERROR;
+  var friendlyError = retrieveFriendlyError(cartoError);
+  track(friendlyError);
+
+  friendlyError.name = 'CartoError';
+  friendlyError.stack = (new Error()).stack;
+
+  return friendlyError;
 }
 
+// Windshaft should have been parsed already 
 function _isWindshaftError (error) {
-  return error && error.errors_with_context;
+  return error && error.origin === 'windshaft';
+}
+
+function transformWindshaftError (error, layers, analysis) {
+  var cartoError = {
+    message: error.message,
+    type: error.type,
+    origin: error.origin
+  };
+  if (error.type === 'layer' && layers) {
+    cartoError.layer = layers.findById(error.layerId);
```

Comment:
Instead of having .layer and .source, perhaps we could find a generic name for the resource/model that failed (like Backbone does) so that it works for both cases. Also, we should document this/these property/ies.

Reply:
That's an option as long as we document the different types that can be returned.
Having separated properties avoid the user to duck type what object it is.